File: Derivatives.py
from typing import Union
import time


class Empty:
    pass


class Epsilon:
    pass


class Char:
    def __init__(self, c):
        self.value = c


class Star:
    def __init__(self, r):
        self.value = r


class Alt:
    def __init__(self, a, b):
        self.a = a
        self.b = b


class Concat:
    def __init__(self, a, b):
        self.a = a
        self.b = b


def match(r, s: str) -> bool:
    return nullable(derivative_(s, r)) is Epsilon


def derivative_(s: str, r):
    derivative_(s[1:], derivative(Char(s[0]), r)) if s else r

# derivative :: Char -> Regexp -> Regexp

def derivative(a: Char, r: Union[Empty, Epsilon]):
    return Empty()


def derivative(a: Char, r: Char):
    if a.value == r.value:
        return Epsilon()
    else:
        return Empty()


def derivative(a: Char, r: Alt):
    return Alt(derivative(a, r.a), derivative(a, r.b))


def derivative(a: Char, r: Star):
    return Concat(derivative(a, r.value), r)


def derivative(a: Char, r: Concat):
    return Alt(Concat(derivative(a, r.a), r.b), Concat(nullable(r.a), derivative(a, r.b)))


# intersect :: Regexp -> Regexp -> Regexp
def intersect(em: Empty, other):
    return Empty()


def intersect(other, em: Empty):
    return Empty()


def intersect(a: Epsilon, b: Epsilon):
    return Epsilon()


# union :: Regexp -> Regexp -> Regexp
def union(eps: Epsilon, other):
    return Epsilon()


def union(other, eps: Epsilon):
    return Epsilon()


def union(a: Empty, b: Empty):
    return Empty()


# nullable :: Regexp -> Regexp
def nullable(r: Union[Empty, Char]):
    return Empty()


def nullable(eps: Epsilon):
    return Epsilon()


def nullable(c: Concat):
    return intersect(nullable(c.a), nullable(c.b))


def nullable(a: Alt):
    return union(nullable(a.a), nullable(a.b))


def nullable(s: Star):
    return Epsilon()

File: Derivatives.hs
module Regexp where

import Prelude hiding (seq)

data Regexp = Empty
              | Epsilon
              | Char Char
              | Star Regexp
              | Alt Regexp Regexp
              | Concat Regexp Regexp
            deriving (Eq, Ord)

match :: Regexp -> String -> Bool
match r s = nullable (derivative' s r) == Epsilon

derivative :: Char -> Regexp -> Regexp
derivative a Empty = Empty 
derivative a Epsilon = Empty
derivative a (Char b) | a == b = Epsilon 
derivative a (Char b) = Empty
derivative a (Alt s t) = Alt (derivative a s) (derivative a t)
derivative a (Star s) = Concat (derivative a s) (Star s) 
derivative a (Concat s t) = Alt (Concat (derivative a s) t) (Concat (nullable s) (derivative a t))

intersect :: Regexp -> Regexp -> Regexp
intersect Empty _ = Empty 
intersect _ Empty = Empty 
intersect Epsilon Epsilon = Epsilon
intersect _ _ = error "Sorry, intersect is unimplemented for these Regexp types"

union :: Regexp -> Regexp -> Regexp
union Epsilon _ = Epsilon 
union _ Epsilon = Epsilon  
union Empty Empty = Empty
union _ _ = error "Sorry, intersect is unimplemented for these Regexp types"

nullable :: Regexp -> Regexp
nullable Empty = Empty
nullable Epsilon = Epsilon
nullable (Char a) = Empty
nullable (Concat s t) = intersect (nullable s) (nullable t)
nullable (Alt s t) = union (nullable s) (nullable t)
nullable (Star s) = Epsilon

derivative' :: String -> Regexp -> Regexp
derivative' (c:cs) r = derivative' cs (derivative c r)
derivative' [] r = r

r1 = Concat (Star (Concat (Char 'a') (Char 'b')) ) (Char 'a')
